src/core/lib/transport/metadata.cc

```diff
@@ -554,3 +444,49 @@ bool grpc_mdelem_eq(grpc_mdelem a, grpc_mdelem b) {
   return grpc_slice_eq(GRPC_MDKEY(a), GRPC_MDKEY(b)) &&
          grpc_slice_eq(GRPC_MDVALUE(a), GRPC_MDVALUE(b));
 }
+
+static void note_disposed_interned_metadata(uint32_t hash) {
+  mdtab_shard* shard = &g_shards[SHARD_IDX(hash)];
+  gpr_atm_no_barrier_fetch_add(&shard->free_estimate, 1);
+}
+
+#ifndef NDEBUG
+void grpc_mdelem_do_unref(grpc_mdelem gmd, const char* file, int line) {
+#else
+void grpc_mdelem_do_unref(grpc_mdelem gmd) {
+#endif
+  switch (GRPC_MDELEM_STORAGE(gmd)) {
+    case GRPC_MDELEM_STORAGE_EXTERNAL:
+    case GRPC_MDELEM_STORAGE_STATIC:
+      return;
+    case GRPC_MDELEM_STORAGE_INTERNED: {
+      interned_metadata* md =
+          reinterpret_cast<interned_metadata*> GRPC_MDELEM_DATA(gmd);
+#ifndef NDEBUG
+      grpc_mdelem_trace_unref(md, md->key, md->value, &md->refcnt, file, line);
+#endif  // ifndef NDEBUG
+      uint32_t hash = md->hash;
+      const gpr_atm prev_refcount = gpr_atm_full_fetch_add(&md->refcnt, -1);
+      GPR_DEBUG_ASSERT(prev_refcount >= 1);
+      if (1 == prev_refcount) {
+        /* once the refcount hits zero, some other thread can come along and
+           free md at any time: it's unsafe from this point on to access it */
+        note_disposed_interned_metadata(hash);
+      }
+      break;
+    }
+    case GRPC_MDELEM_STORAGE_ALLOCATED: {
+      allocated_metadata* md =
+          reinterpret_cast<allocated_metadata*> GRPC_MDELEM_DATA(gmd);
+#ifndef NDEBUG
+      grpc_mdelem_trace_unref(md, md->key, md->value, &md->refcnt, file, line);
+#endif  // ifndef NDEBUG
+      const gpr_atm prev_refcount = gpr_atm_full_fetch_add(&md->refcnt, -1);
+      GPR_DEBUG_ASSERT(prev_refcount >= 1);
+      if (1 == prev_refcount) {
+        md->~allocated_metadata();
```

Comment:
We are not using the CTOR of allocated_metadata. I'd suggest converting this structure to C++, adding proper ctor and using grpc_core::New() and Delete() instead.

Reply:
DOne.

Reply:
Done.